rpc: use sync.Pools for snappy readers and writers

petermattis opened the pull request:
Creating a snappy.{Reader,Writer} on every compression/decompression is
fairly expensive. This reduced the performance hit for enabling
compression in a local 3-node cluster from 38% to 3%.

See #14721

petermattis committed:
rpc: use sync.Pools for snappy readers and writers

Creating a snappy.{Reader,Writer} on every compression/decompression is
fairly expensive. This reduced the performance hit for enabling
compression in a local 3-node cluster from 38% to 3%.

See #14721

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15826)
<!-- Reviewable:end -->

tschottdorf commented:
Fantastic, LGTM

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Nice, I wasn't expecting this to account for nearly the whole performance difference.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/rpc/snappy.go, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15826#-KjjhFcTKuMGZdA65sLd:-KjjhFcTKuMGZdA65sLe:bxxq3g5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5794eb6bd070c2cb0584aa126434df74c7d8827a/pkg/rpc/snappy.go#L28)):*
> ```Go
> // safe as multiple goroutines may be using the same compressor/decompressor
> // for different streams on the same connection.
> var snappyWriterPool = sync.Pool{}
> ```

These can be `var snappyWriterPool sync.Pool` (and I think this style is more typical for us)

---

*[pkg/rpc/snappy.go, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15826#-KjjhMf-8fi_JU2hkYm6:-KjjhMf-8fi_JU2hkYm7:bjcviz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5794eb6bd070c2cb0584aa126434df74c7d8827a/pkg/rpc/snappy.go#L37)):*
> ```Go
> 	z, ok := snappyWriterPool.Get().(*snappy.Writer)
> 	if !ok {
> 		z = snappy.NewWriter(w)
> ```

This also changes from `NewBufferedWriter` to `NewWriter`. Did this make a difference too? (I see that NewBufferedWriter only helps when you make more than one `Write` call, so the change is appropriate here)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15826#-:-KjjhiZvd02FooToMECk:b-nl8fyh)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15826#-:-Kjjlwn5reYdOtK-1rU1:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
@bdarnell I wasn't expecting this to account for the performance difference either. Btw, I also tried a no-op compressor and that showed zero slowdown vs no compression.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/rpc/snappy.go, line 28 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15826#-KjjhFcTKuMGZdA65sLd:-KjjsWYGW2CXye2D_a6V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5794eb6bd070c2cb0584aa126434df74c7d8827a/pkg/rpc/snappy.go#L28)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

These can be `var snappyWriterPool sync.Pool` (and I think this style is more typical for us)
</blockquote></details>

Done.

---

*[pkg/rpc/snappy.go, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15826#-KjjhMf-8fi_JU2hkYm6:-KjjsX6vcFf9Mm2tOll7:b-3ob409) ([raw file](https://github.com/cockroachdb/cockroach/blob/5794eb6bd070c2cb0584aa126434df74c7d8827a/pkg/rpc/snappy.go#L37)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This also changes from `NewBufferedWriter` to `NewWriter`. Did this make a difference too? (I see that NewBufferedWriter only helps when you make more than one `Write` call, so the change is appropriate here)
</blockquote></details>

Ah, I reverted back to using `NewBufferedWriter`. My initial stab at this was along the lines of your suggestion to embed the writer in `snappyCompressor`. But then testing showed corruption. I initially suspected something fouled up in my usage and switched to using `NewWriter` and just never switched it back. There is no performance difference between `NewWriter` and `NewBufferedWriter` in local tests. I'm going to test on `denim` shortly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15826#-:-KjjswQyifmbfXvGT1nX:bfaf1u4)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Another thought I had on the way home is that there might be benefit to explicitly compressing the `WriteBatch` inside of a `RaftCommand`. Leaving the compression at the RPC level means that we end up compressing a the `WriteBatch` once for each follower, then decompressing it on the follower in order to write it to the Raft log and then letting RocksDB compress it again. By compressing the `WriteBatch` we would reduce the number of compressions by 1 and also reduce the in-memory size of the Raft entry cache. The downside to this is that there are backward compatibility issues that I'm not sure how to address (yet). 

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15826#-:-KjjtGtB_rcAHUXSf7_8:bu3prlh)*
<!-- Sent from Reviewable.io -->

tamird commented:
On Tue, May 9, 2017 at 9:10 PM, Peter Mattis <notifications@github.com>
wrote:

> By compressing the WriteBatch we would reduce the number of compressions
> by 1
>

By N-2, right?

petermattis commented:
> By N-2, right?

Yes.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15826#-:-KjjyfTbTqXTcKRRJzVJ:b-h749v5)*
<!-- Sent from Reviewable.io -->

petermattis commented:
The numbers below were gathered from a 3-node cluster running on `denim`. The before numbers (master) are similar to the previously reported 40% slowdown (38% in the 95% read case and 31% in the 0% read case). The after numbers show there is no appreciable performance difference for enabling compression. 

```
kv --read-percent=95 --splits=1000 --concurrency=100 --duration=2m
master(no-comression):
_elapsed___errors____________ops___ops/sec(cum)__p95(ms)__p99(ms)_pMax(ms)
  120.0s        0        2708871        22569.1     14.7     35.7   7247.8
master(snappy):
_elapsed___errors____________ops___ops/sec(cum)__p95(ms)__p99(ms)_pMax(ms)
  120.0s        0        1660698        13838.1     23.1     62.9    671.1
15826(snappy):
_elapsed___errors____________ops___ops/sec(cum)__p95(ms)__p99(ms)_pMax(ms)
  120.0s        0        2719538        22661.7     15.2     39.8    570.4

kv --read-percent=0 --splits=1000 --concurrency=100 --duration=2m
master(no compression):
_elapsed___errors____________ops___ops/sec(cum)__p95(ms)__p99(ms)_pMax(ms)
  120.0s        0         691862         5765.1     31.5     56.6   3623.9
master(snappy):
_elapsed___errors____________ops___ops/sec(cum)__p95(ms)__p99(ms)_pMax(ms)
  120.0s        0         475428         3961.8     46.1     58.7   3758.1
15826(snappy):
_elapsed___errors____________ops___ops/sec(cum)__p95(ms)__p99(ms)_pMax(ms)
  120.0s        0         687404         5727.3     33.6     56.6   6710.9
```

petermattis commented:
Small follow-up, I also tried using `NewWriter` instead of `NewBufferedWriter` and saw zero performance difference. Given that `NewWriter` is marked as deprecated and we'd need to special case using it for our linter checks, I'm going to stick with `NewBufferedWriter`.

a-robinson commented:
Well this was a nice little chunk of code. <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15826#-:-KjmeClbCa8sdRigabQo:b-kqii7j)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Btw, I'm pretty sure the benefit is due to the massive decrease in allocated memory for every RPC which would have added significant GC pressure. A `snappy.{Reader,Writer}` allocates 2 ~64KB buffers. Before using the `sync.Pool`, those allocations were happening for every RPC, no matter how small. Ouch!

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15826#-:-Kjmebusm2NFzByXsiRJ:bd2l0u6)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Heh: https://github.com/grpc/grpc-go/pull/1217
